### PR TITLE
vLLM example with mistral7B model

### PR DESCRIPTION
**Pull request by agunapal**
## Description

This example shows how to use vLLM to run inference on mistral7b model

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

```
curl -v "http://localhost:8080/predictions/mistral7b" -T sample_text.txt
```

results in the following output
```
Mayonnaise is made of eggs, oil, vinegar, salt and pepper. Using an electric blender, combine all the ingredients and beat at high speed for 4 to 5 minutes.

Try it with some mustard and paprika mixed in, and a bit of sweetener if you like. But use real mayonnaise or it isn’t the same. Marlou

What in the world is mayonnaise?
```

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: examples/large_models/vllm/mistral/model-config.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+maxBatchDelay: 100
+responseTimeout: 1200
+deviceType: "gpu"
+
+handler:
+    model_name: "mistralai/Mistral-7B-v0.1"
+    model_path: "/home/ubuntu/serve/examples/large_models/vllm/mistral/model/models--mistralai--Mistral-7B-v0.1/snapshots/5e9c98b96d071dce59368012254c55b0ec6f8658"
+    max_new_tokens: 100
+    manual_seed: 40
+    fast_kernels: True
```
Comment (lxning) [inline examples/large_models/vllm/mistral/model-config.yaml:13]: Can you demo using multiple gpus in this example?

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            pertaining to the model artifacts parameters.
+        """
+        model_dir = ctx.system_properties.get("model_dir")
+        self.max_new_tokens = int(ctx.model_yaml_config["handler"]["max_new_tokens"])
+        model_name = ctx.model_yaml_config["handler"]["model_name"]
+        model_path = ctx.model_yaml_config["handler"]["model_path"]
+        seed = int(ctx.model_yaml_config["handler"]["manual_seed"])
+        torch.manual_seed(seed)
+
+        self.model = LLM(model=model_path)
```
Comment (lxning) [inline examples/large_models/vllm/mistral/custom_handler.py:40]: LLM has internal worker orchestration based on ray. it is inefficient to directly on LLM. Can you demo deploying  a worker of a model to multiple gpus and reduce the overhead as possible as you can?

File: examples/large_models/vllm/mistral/model-config.yaml (hunk 1 above)
Comment (agunapal) [reply examples/large_models/vllm/mistral/model-config.yaml]: Thanks for the feedback. Updated the example

File: examples/large_models/vllm/mistral/custom_handler.py (hunk 2 above)
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Updated the example.

File: examples/large_models/vllm/mistral/model-config.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+torchrun:
+    nproc-per-node: 4
+
+handler:
+    model_name: "mistralai/Mistral-7B-v0.1"
+    model_path: "/home/ubuntu/serve/examples/large_models/vllm/mistral/model/models--mistralai--Mistral-7B-v0.1/snapshots/5e9c98b96d071dce59368012254c55b0ec6f8658"
+    max_new_tokens: 100
+    manual_seed: 40
+    tensor_parallel_size : 4
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/model-config.yaml]: You are right. Not sure if there is a valid usecase where they are different. Removed it

File: examples/large_models/vllm/mistral/Readme.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+python ../../Huggingface_accelerate/Download_model.py --model_path model --model_name mistralai/Mistral-7B-v0.1
+```
+Model will be saved in the following path, `mistralai/Mistral-7B-v0.1`.
+
+### Step 2: Generate MAR file
+
+Add the downloaded path to " model_path:" in `model-config.yaml` and run the following.
+
+```bash
+torch-model-archiver --model-name mistral7b --version 1.0 --handler custom_handler.py --config-file model-config.yaml -r requirements.txt --archive-format tgz
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/Readme.md:25]: Good point. Removed it.

File: examples/large_models/vllm/mistral/Readme.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+### Step 3: Add the mar file to model store
+
+```bash
+mkdir model_store
+mv mistral7b.tar.gz model_store
+```
+
+### Step 3: Start torchserve
+
+Update config.properties and start torchserve
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/Readme.md]: Removed it.

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    def __init__(self):
+        super(LlamaHandler, self).__init__()
+        self.max_new_tokens = None
+        self.tokenizer = None
+        self.initialized = False
+
+    def initialize(self, ctx: Context):
+        """In this initialize function, the HF large model is loaded and
+        partitioned using DeepSpeed.
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Fixed all the copy paste issues :D

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        Returns:
+            tuple: A tuple with two tensors: the batch of input ids and the batch of
+                attention masks.
+        """
+        input_texts = [data.get("data") or data.get("body") for data in requests]
+        # return torch.as_tensor(input_texts, device=self.device)
+        input_texts = [
+            input_text.decode("utf-8")
+            for input_text in input_texts
+            if isinstance(input_text, (bytes, bytearray))
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py:57]: Good catch. Added an else condition

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        ]
+        return input_texts
+
+    def inference(self, input_batch):
+        """
+        Predicts the class (or classes) of the received text using the serialized transformers
+        checkpoint.
+        Args:
+            input_batch (tuple): A tuple with two tensors: the batch of input ids and the batch
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Fixed all the copy paste issues :D

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        return outputs
+
+    def postprocess(self, inference_output):
+        """Post Process Function converts the predicted response into Torchserve readable format.
+        Args:
+            inference_output (list): It contains the predicted response of the input text.
+        Returns:
+            (list): Returns a list of the Predictions and Explanations.
+        """
+        return [inference_output[0].outputs[0].text]
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Good catch. Fixed the batching issue. Verified it

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+from vllm import LLM, SamplingParams
+
+from ts.context import Context
+from ts.torch_handler.base_handler import BaseHandler
+
+logger = logging.getLogger(__name__)
+logger.info("vLLM version %s", vllm.__version__)
+
+
+class LlamaHandler(BaseHandler, ABC):
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Done

File: examples/large_models/vllm/mistral/custom_handler.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        partitioned using DeepSpeed.
+        Args:
+            ctx (context): It is a JSON Object containing information
+            pertaining to the model artifacts parameters.
+        """
+        model_dir = ctx.system_properties.get("model_dir")
+        self.max_new_tokens = int(ctx.model_yaml_config["handler"]["max_new_tokens"])
+        model_name = ctx.model_yaml_config["handler"]["model_name"]
+        model_path = ctx.model_yaml_config["handler"]["model_path"]
+        tp_size = ctx.model_yaml_config["handler"]["tensor_parallel_size"]
```
[the remark this answers is not in the document]
Comment (agunapal) [reply examples/large_models/vllm/mistral/custom_handler.py]: Done

**Review by mreso — APPROVED** [review]
Thanks for addressing all comments! LGTM

**Pull request merged**